// @reef-defi/hardhat-reef/dist/src/proxies/signers/EthersSigner.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.EthersSigner = void 0;
const ethers_1 = require("ethers");
// Class is never used and is defined just for type testing purposes
// This implementation should be as follows but SignerWithAddress is a private class. This code is just his copy.
// export class ES extends SignerWithAddress implements ReefSigner {}
class EthersSigner extends ethers_1.Signer {
    constructor(address, _signer) {
        super();
        this.address = address;
        this._signer = _signer;
        this.provider = _signer.provider;
    }
    static async create(signer) {
        return new EthersSigner(await signer.getAddress(), signer);
    }
    async getAddress() {
        return this.address;
    }
    signMessage(message) {
        return this._signer.signMessage(message);
    }
    signTransaction(transaction) {
        return this._signer.signTransaction(transaction);
    }
    sendTransaction(transaction) {
        return this._signer.sendTransaction(transaction);
    }
    connect(provider) {
        return new EthersSigner(this.address, this._signer.connect(provider));
    }
    _signTypedData(...params) {
        return this._signer._signTypedData(...params);
    }
    toJSON() {
        return `<SignerWithAddress ${this.address}>`;
    }
    async claimDefaultAccount() { }
}
exports.EthersSigner = EthersSigner;
